url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/security/SeedsGenerater.java

```diff
@@ -0,0 +1,164 @@
+package utility.security;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Random;
+
+public class SeedsGenerater {
+	public static BigInteger[] seeds, rsa_g_seeds, g_seeds;
+	public static Paillier paillier = new Paillier(true);
+	final static BigInteger ONE = BigInteger.ONE;
+	final static BigInteger TWO = BigInteger.ONE.add(BigInteger.ONE);
+	public static RSA rsa = new RSA();
+	public static int base = 16; 
+	public String filename = "./seeds";
+	public SeedsGenerater(){}
+	public SeedsGenerater(int bitlength){
+		seeds = new BigInteger[bitlength + 1];
+		rsa_g_seeds = new BigInteger[bitlength + 1];
+		g_seeds = new BigInteger[bitlength + 1];
+		seeds[0] = BigIntegerUtility.ZERO;
+		g_seeds[0] = paillier.Encryption(seeds[0], BigIntegerUtility.ONE);
+		rsa_g_seeds[0] = rsa.encrypt(g_seeds[0]);
+		//g_seeds[0] = g_seeds[0].multiply(paillier.r2n).mod(paillier.nsquare);
+		BigInteger val = BigIntegerUtility.ONE;
+		for(int i = 0; i < bitlength; i++){
+			seeds[i + 1] = val;
+			g_seeds[i + 1] = paillier.Encryption(val, BigIntegerUtility.ONE);
+			rsa_g_seeds[i + 1] = rsa.encrypt(g_seeds[i + 1]);
+			//g_seeds[i + 1] = g_seeds[i + 1].multiply(paillier.r2n).mod(paillier.nsquare);
+			val = val.multiply(BigInteger.valueOf(base));
+		}
+		try {
+			DataOutputStream seedsdos = new DataOutputStream(new FileOutputStream(filename + ".dat"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *seedsdos*. The resource is referenced at lines 40, 43, 44, 46, 54. The resource is closed at line 55. There are other execution paths that do not contain closure statements, for example, when DataOutputStream.flush throws an exception. Ensure *seedsdos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).